app/src/main/java/programatorus/client/shared/FirmwareService.kt

```diff
@@ -1,22 +1,33 @@
 package programatorus.client.shared
 
+import programatorus.client.device.IDevice
 import programatorus.client.model.Firmware
+import programatorus.client.model.FirmwareData
+import java.util.concurrent.CompletableFuture
 
 class FirmwareService() {
     val repository = FavoritesRepository<Firmware>()
+    lateinit var client: IDevice
 
-    fun pull() {
-        // TODO: MSG Get Firmware files
-        val all = (1..5).map { Firmware(it.toString(), true) } + (6..12).map { Firmware(it.toString(), false) }
-        val favorites = (1..5).map { Firmware(it.toString(), true) }
-
-        repository.setState(all, favorites)
+    fun pull(): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+        client.getFirmware().thenAccept { firmware ->
+            repository.setState(firmware.all, firmware.favorites)
+            future.complete(Unit)
+        }
+        return future
     }
 
-    fun push() {
-        repository.getAll()
-        repository.getFavorites()
+    fun push(): CompletableFuture<Boolean> {
+        val future = CompletableFuture<Boolean>()
 
-        // TODO: MSG Post Firmware files
+        client.putFirmware(getFirmwareData())
```

Comment:
Is failure dropped here intentionally?

Reply:
It is intentional. It was meant for cases when file upload/delete would make data irrelevant. We were talking about awaiting result after file upload - it would fix issue